src/functional/merge.js

```diff
@@ -68,7 +69,11 @@ export function mergeWithSources(collection, sources, merger) {
         }
       };
   for (let i = 0; i < sources.length; i++) {
-    Collection(sources[i]).forEach(mergeItem);
+    const source = sources[i];
+    if (isArray && isKeyed(source)) {
```

Comment:
It looks to me like you're missing the reciprocal case: where the left hand side is keyed and the right hand is not. You could instead write `isArray === isKeyed(source)`. It's a little tricky, but it has the correct behavior.

Reply:
"a little tricky" is the foundation of "WTF" months/years later. I would generally prefer readability in our changes, since the original code has enough fancy short-written cleverness 😅

Reply:
Yeah, the different cases where it throws an error are somewhat spread out across the code-base, but I believe all the cases are covered in one place or another. I was going for the most minimal impact change, but perhaps a more centralized check is called for (and more easy to follow). There also doesn't seem to be much validation throughout the library, so I was trying to keep the validation to a minimum.

The current check covers the (array, Map) case. The suggested change would additionally cover the (object, array) and (object, List) case which are checked for later in the code. However it would erroneously fail for (object, object) (since `isKeyed(object)` is `false`).

A more comprehensive check might look something like this:
```
  const isArray = Array.isArray(collection);
  const isObject = !isArray;
  for (let i = 0; i < sources.length; i++) {
    const source = sources[i];
    if (
      (isArray && (isPlainObject(source) || isKeyed(source))) ||
      (isObject && (Array.isArray(source) || isIndexed(source)))
    ) {
      throw new TypeError('Expected sources to match: ' + source);
    }
    Collection(source).forEach(mergeItem);
  }
```

I thought about moving the validation against adding a key-value pair to an indexed collection deeper into the code (like it is for `KeyedCollection`), but I'm not sure it's possible since it seems reasonable to want to make an indexed collection out of key-value pairs. So perhaps the correct answer is the more comprehensive check. Does the above check look good?

Reply:
I still think it'd be smart to use `===` or `!==` for speed. Then you end up writing something like `sourceKeyed !== collKeyed`.

Reply:
Yeah, that's fine (although I tend to think it's premature optimization), however I don't think this can be reduced to a single comparison because `isPlainObject(source) || isKeyed(source)` is not the opposite of `Array.isArray(source) || isIndexed(source)` so there's no way to have a `sourceKeyed` variable to capture what I'm trying to express above.

Reply:
I've updated the type check with the code shown above.